File: utils/resolveVersionConflict.py
# This script programmatically resolves merge conflicts in a text file.

# The conflict resolution logic goes here.
# Should return the replacement_string, or None if you don't wish to resolve this conflict.
def resolve_conflict(head, theirs, entire_conflict):
    if head.find('"version":') > 0:
        return head
    else:
        return entire_conflict

# ----------------------------------------

import re
import sys

if __name__ == '__main__':

    usage = '''
    Usage:
    First argument is the path of the source file.
    Second argument, if specified, should be "overwrite" and means that the source file will be overwritten. (Use with care)
    '''

    if not (2 <= len(sys.argv) <= 3):
        print (usage)
        sys.exit('Bad number of args.')

    do_overwrite_source_file = False
    if len(sys.argv) == 3:
        if sys.argv[2].lower() == 'overwrite':
            do_overwrite_source_file = True
        else:
            sys.exit('Last argument should be "overwrite" if specified.')

    source_filename = sys.argv[1]
    with open(source_filename) as source_file:
        source_text = source_file.read()

    pattern = re.compile(
    '''\\n<<<<<<< HEAD'''
    '''(?P<head>.*?)'''
    '''\\n======='''
    '''(?P<theirs>.*?)'''
    '''\\n>>>>>>> (?P<branch_name>[^\\n]+)''', re.DOTALL)

    resolved_count = 0
    total_count = 0
    replacements = [] # List of (start_index, end_index, replacement_string) tuples.

    for match in re.finditer(pattern, source_text):
        head   = match.groupdict()['head']
        theirs = match.groupdict()['theirs']
        match_string = match.string[match.start() : match.end()]
        replacement_string = resolve_conflict(head, theirs, match_string)
        if replacement_string is not None:
            replacements.append((match.start(), match.end(), replacement_string))
            resolved_count += 1
        total_count += 1
        print('branch_name:       ', repr(match.groupdict()['branch_name']))
        print('head:              ', repr(head))
        print('theirs:            ', repr(theirs))
        print('match_string:      ', repr(match_string))
        print('replacement_string:', repr(replacement_string))
        print()
        print('--------------------')
        print()

    # Perform the replacements in reverse, so as not to mess up the indices.
    result_text = source_text
    replacements.sort(reverse=True)
    for (start, end, replacement_string) in replacements:
        result_text = result_text[:start] + replacement_string + result_text[end:]

    if do_overwrite_source_file:
        print('Overwriting file.')
        with open(source_filename, 'w') as source_file:
            source_file.write(result_text)
    else:
        print('Dry run (Not overwriting file).')

    print('Resolved %d out of %d conflicts.' % (resolved_count, total_count))

File: utils/resolveVersionConflict.js
// This script programmatically resolves merge conflicts in a text file.

const { readFileSync, writeFileSync } = require('fs');

const [, , file, overwrite] = process.argv;

// The conflict resolution logic goes here.
// Should return the replacementString, or null if you don't wish to resolve this conflict.
function resolveConflict(head) {
  if (!head.includes('"version":') || !head.includes('projectVersion=')) {
    throw new Error("Found non-version related conflict, exiting auto resolution")
  }
  return head;
}

const sourceText = readFileSync(file).toString();

const lend = `(\n|\r\n)`; // Support cross-platform line endings
const pattern = new RegExp(
  `<<<<<<< HEAD${lend}(?<head>.*?)${lend}=======${lend}(?<theirs>.*?)${lend}>>>>>>> (?<branchName>[^${lend}]+)$`,
  'gsm'
);

let resultText;
let resolvedCount = 0;
let totalCount = 0;
const replacements = []; // List of (start_index, end_index, replacement_string) tuples.

/* eslint-disable-next-line no-restricted-syntax */
for (const match of sourceText.matchAll(pattern)) {
  console.log(match[0]);
  const { head, theirs, branchName } = match.groups;
  const [matchString] = match;
  const replacementString = resolveConflict(head, theirs, matchString)
  if (replacementString) {
    replacements.push([match.index, match.index + matchString.length, replacementString])
    resolvedCount += 1
  }
  totalCount += 1
  console.log('branchName:      ', branchName)
  console.log('head:            ', head)
  console.log('theirs:          ', theirs)
  console.log('matchString:     ', matchString)
  console.log('replacementString:', replacementString)
  console.log('--------------------')

  resultText = sourceText.replace(matchString, replacementString);
}

if (replacements.length > 1) {
  throw new Error('Found multiple conflicts, exiting auto resolution');
}

if (overwrite) {
  console.log('Overwriting file.')
  writeFileSync(file, resultText)
} else {
  console.log('Dry run (Not overwriting file).');
  console.log(resultText);
}

console.log('Resolved %d out of %d conflicts.', resolvedCount, totalCount);
